### DISCUSSION TITLE
convert command ignoring memory limit set

### DISCUSSION THREAD

**Discussion by ankur8** [ImageMagick/ImageMagick#3363] [General]
Hi Team,

I have tried setting memory request limit , map limit, memory limit but convert command in case of jpeg and png is ignoring same and making my rss and res and working memory set cross the max limit assigned in pod. I can not go on increasing my memory for jpeg and png image type for my pod in k8s cluster.

I am running imagemagick ver 7. Could you shed some help here. ?

os - ubuntu 18

**Comment by snibgo** [comment]
The ImageMagick resource limits, on memory etc, apply only to ImageMagick code. They don't apply to libraries that IM uses, eg to read and write PNG files, or to external delegates such as Ghostscript that rasterizes PDF files.

This may be the problem you are encountering.

You can set IM limits so low that IM doesn't use any significant memory for itself. This may solve your memory problem, but at a cost to performance.

**Comment by urban-warrior** [comment]
To reduce the memory footprint in ImageMagick, try these command-line options:

```
magick -limit area 0 -limit thread 1 ...
```
Performance will suffer as the image pixels are shuffled to disk rather than memory and you reduce computation to a single core.  As @snibgo points out, ImageMagick does not control the memory that delegate libraries consume (e.g. libjpeg, libtiff, etc.).  And in many cases, ImageMagick image processing algorithms require certain memory requirements to complete.  The only other options to reduce the memory footprint is to build a non-HDRI release of ImageMagick (2 bytes per pixel rather than 4), reduce the image size, utilize a distributed pixel cache, or use some other utility that may have a reduced memory footprint. The distributed pixel cache offloads pixels to a remote server that presumably has more available memory, but that shuffles pixels over the network which further reduces performance.

You can set the `system:max-memory-request` policy to gracefully exit if ImageMagick exceeds the imposed memory limit.

**Comment by ankur8** [comment]
@snibgo @urban-warrior 

Some more insights for more inputs as below 

imagemagick version

root@media-delivery-api-8d9b6f6c6-xp5w2:/opt/flare# identify -version
Version: ImageMagick 7.0.10-29 Q16 x86_64 2020-09-07 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(4.5) 
We are converting tiff to jpeg or png . Let me show you exact  commands. I tried all below except thread no memory or map request is getting considered by image magick convert command for image conversion. 

policy domain="resource" name="disk" value="4GiB"/>
policy domain="resource" name="area" value="500MiB"/>' /usr/local/etc/ImageMagick-7/policy.xml
policy domain="system" name="max-memory-request" value="500MiB"/> /usr/local/etc/ImageMagick-7/policy.xml
policy domain="resource" name="memory" value="500MiB"/>' /usr/local/etc/ImageMagick-7/policy.xml
policy domain="resource" name="map" value="500MiB"/>/usr/local/etc/ImageMagick-7/policy.xml

Below is only respected.
policy domain="resource" name="thread" value="2"/>#' /usr/local/etc/ImageMagick-7/policy.xml

==========

if you all see this is the output of top command and rss used by convert
 1323.1m **1.166g**   7.5m R 40.5  3.7   0:01.22 **convert -resize 200x211! +repage** /appl/media/images-v2/nce-sformation/a44e2dsfasfs-a042-4531-a488-bcd613c7b723/media.jpeg /appl/media/tmp/folder1/

 1089.8m 945.5m   6.3m R 68.4  2.9   0:02.06 convert -resize 200x211! +repage /appl/media/images-v2/flare-finance-transformation/a44ea1b4-a042-4531-a488-bcd613c7b723/media.jpeg /appl/me


![image](https://user-images.githubusercontent.com/7699839/110228133-b78db600-7f24-11eb-818e-a7b84d2499c7.png)



look no limit i can define on ram consumption on container :-( Bringing it up if in imagmagic we can solve before moving to another alternative.

**Reply by urban-warrior** [reply]
Verify your policy is defined as expected with `magick -list policy`.  Try setting the area limit to 0.  You could try `-scale` rather than `-resize`, `-scale` consumes less memory.  Otherwise we are at the point of diminishing returns.  Increase the memory limits of your container or adopt an alternative solution.

**Comment by snibgo** [comment]
@ankur8: I suggest you also try "magick -list resource". Please show the exact text output.

> We are converting tiff to jpeg or png . Let me show you exact commands.

You don't show exact commands. For example, you show...
```
convert -resize 200x211! +repage /appl/media/images-v2/nce-sformation/a44e2dsfasfs-a042-4531-a488-bcd613c7b723/media.jpeg /appl/media/tmp/folder1/
```
... which is not a complete command, as there is no output filename.

You use IM v7. I suggest using "magick" instead of "convert". The command should also follow the logical order: read the image, then process it (eg "-resize"), when write it.

**Reply by ankur8** [reply]
@snibgo 

magick -list resource
root@media-delivery-api-8d9b6f6c6-7xkrk:/opt/flare# magick -list resource
Resource limits:
  Width: 107.374MP
  Height: 107.374MP
  Area: 67.351GP
  List length: unlimited
  Memory: 31.3628GiB
  Map: 62.7255GiB
  Disk: unlimited
  File: 786432
  Thread: 4
  Throttle: 0
  Time: unlimited

I have set given 7gb ram it looks need to increase more. Also threading limit working fine i can see only max 4 threads convert is running with in top output. 

output filename  is in jpeg only kind of resizing(-resize) and with -format commands we are converting from tiff to jpeg. I take your inputs as improvement i.e magick instead of convert for resize use. 

Switching to another tool kinda rework and testing. I am expecting if we can improve this only. I could not see any bug in debain/sid as we are using 5.4.0-1040-azure #42~18.04.1-Ubuntu release or in libjpeg.so dynamic linked lib we are using.

**Reply by ankur8** [reply]
we are using Version: ImageMagick 7.0.10-29 Q16 x86_64 2020-09-07 https://imagemagick.org . should i upgrade to 7.1 ? @snibgo

**Comment by snibgo** [comment]
You say:
> I have tried setting memory request limit , map limit, memory limit but convert command in case of jpeg and png is ignoring same ...

... but your "magick -list resource" shows no significant limits.

How did you try to set the limits?

**Reply by ankur8** [reply]
Ya I set that up  with this change in policy.xml but at the moment and it was showing all those limit for area, memory etc.  But for now I have removed all those except thread.
policy domain="resource" name="disk" value="4GiB"/>
policy domain="resource" name="area" value="500MiB"/>' /usr/local/etc/ImageMagick-7/policy.xml
policy domain="system" name="max-memory-request" value="500MiB"/> /usr/local/etc/ImageMagick-7/policy.xml
policy domain="resource" name="memory" value="500MiB"/>' /usr/local/etc/ImageMagick-7/policy.xml
policy domain="resource" name="map" value="500MiB"/>/usr/local/etc/ImageMagick-7/policy.xml
